stellar_rust_sdk/src/trades/all_trades_request.rs

```diff
@@ -0,0 +1,179 @@
+            base_asset: Some(BaseAsset(base_asset)),
+            ..self
+        })
+    }
+
+    /// Specifies the counter asset in the request.
+    ///
+    /// # Arguments
+    ///
+    /// * `counter_asset` - The counter asset type to filter the trades. It can be one of the following:
+    ///     - `AssetType::Native`
+    ///     - `AssetType::Alphanumeric4(AssetData)`
+    ///     - `AssetType::Alphanumeric12(AssetData)`
+    ///
+    /// # Returns
+    ///
+    /// The updated `AllTradesRequest` with the counter asset set.    
+    pub fn set_counter_asset(
+        self,
+        counter_asset: AssetType,
+    ) -> Result<AllTradesRequest, String> {
+        Ok(AllTradesRequest {
+            counter_asset: Some(CounterAsset(counter_asset)),
+            ..self
+        })
+    }
+}
+
+impl Request for AllTradesRequest {
+    fn get_query_parameters(&self) -> String {
+        let mut query: Vec<String> = Vec::new();
+        
+        if let Some(base_asset) = &self.base_asset {
+            match &base_asset.0 {
+                AssetType::Native => {
+                    query.push(format!("base_asset_type=native"));
+                }
+                AssetType::Alphanumeric4(asset) => {
+                    query.push(format!("base_asset_type=credit_alphanum4"));
+                    query.push(format!("&base_asset_code={}", asset.asset_code));
+                    query.push(format!("&base_asset_issuer={}", asset.asset_issuer));
+                }
+                AssetType::Alphanumeric12(asset) => {
+                    query.push(format!("base_asset_type=credit_alphanum12"));
+                    query.push(format!("&base_asset_code={}", asset.asset_code));
+                    query.push(format!("&base_asset_issuer={}", asset.asset_issuer));
+                }
+            }
+        }
+
+        if let Some(base_asset) = &self.base_asset {
+            match &base_asset.0 {
+                AssetType::Native => {
+                    query.push(format!("&counter_asset_type=native"));
+                }
+                AssetType::Alphanumeric4(asset) => {
+                    query.push(format!("&counter_asset_type=credit_alphanum4"));
+                    query.push(format!("&counter_asset_code={}", asset.asset_code));
+                    query.push(format!("&counter_asset_issuer={}", asset.asset_issuer));
+                }
+                AssetType::Alphanumeric12(asset) => {
+                    query.push(format!("&counter_asset_type=credit_alphanum12"));
+                    query.push(format!("&counter_asset_code={}", asset.asset_code));
+                    query.push(format!("&counter_asset_issuer={}", asset.asset_issuer));
+                }
+            }
+        }
+        query.join("")
+    }
+
+    fn build_url(&self, base_url: &str) -> String {
+        format!(
+            "{}/{}?{}",
+            base_url,
+            super::TRADES_PATH,
+            self.get_query_parameters()
+        )
+    }
+}
```

Comment:
De code voor het ophalen van de QueryString parameters is redelijk duplicaat, met uitzondering can de counter/base prefix. Dit kun je weer samenvoegen in 1 functie. Bijvoorbeeld zoiets:

```rust
impl Request for AllTradesRequest {
    fn add_asset_query(&mut self, query: &mut Vec<String>, asset: &Asset, asset_prefix: &str) {
        let asset_type = match &asset.0 {
            AssetType::Native => "native".to_string(),
            AssetType::Alphanumeric4(asset_data) | AssetType::Alphanumeric12(asset_data) => {
                let asset_type_label = if matches!(asset.0, AssetType::Alphanumeric4(_)) {
                    "credit_alphanum4"
                } else {
                    "credit_alphanum12"
                };
                query.push(format!("&{}_asset_code={}", asset_prefix, asset_data.asset_code));
                query.push(format!("&{}_asset_issuer={}", asset_prefix, asset_data.asset_issuer));
                asset_type_label.to_string()
            }
        };
        query.push(format!("{}_asset_type={}", asset_prefix, asset_type));
    }

    fn get_query_parameters(&self) -> String {
        let mut query = Vec::new();
        if let Some(base_asset) = &self.base_asset {
            self.add_asset_query(&mut query, base_asset, "base");
        }
        if let Some(counter_asset) = &self.counter_asset {
            self.add_asset_query(&mut query, counter_asset, "counter");
        }
        query.join("&").trim_start_matches('&').to_string()
    }

    fn build_url(&self, base_url: &str) -> String {
        format!("{}/{}?{}", base_url, super::TRADES_PATH, self.get_query_parameters())
    }
}
```

Reply:
Dit had ik inderdaad ook gezien. Het gebeurt ook elders in de code, dus misschien kunnen we een stapje verder gaan door er een algemeen iets van te maken?

Reply:
Dat klinkt als een goed idee :D